### PR TITLE
Introduced routines for MPI shared memory

### PR DESCRIPTION
**Pull request by terminationshock**
With this commit, I implemented basic functionality for using MPI shared memory. We can use this concept to reduce the memory usage of an application if large arrays are stored only once per node and not once per MPI task.

See test/test_shared_memory.f90 for a short example.

### CODE REVIEW

File: lib/mpifx_shared_memory.fpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    integer, intent(in) :: length
+    integer, intent(out) :: win
+    ${TYPE}$, pointer, intent(out) :: shared_data(:)
+    integer, intent(out), optional :: error
+
+    integer :: disp_unit, error0, error1
+    integer(MPI_ADDRESS_KIND) :: local_length
+    type(c_ptr) :: baseptr
+
+    disp_unit = kind(shared_data)
```
Comment (aradi) [inline lib/mpifx_shared_memory.fpp:45]:
```suggestion
    disp_unit = storage_size(shared_data) / 8
```
Note: As per Fortran standard, `kind()` returns an arbitrary positive integer number used to enumerate the kinds of a given variable type. For some compilers (intel, gnu) it is accidentally the same, as the number of bytes the given kind occupies, but for some other compilers it is not (e.g. nag returns `1` for single precision and `2` for double precision!).  We should use `storage_size()` instead, which returns the nr. of bits an element of a given type within an array would occupy (including eventual padding in theory).

**Review by aradi — COMMENTED** [review]
Thanks, that's a nice addition. I was not aware of the shared windows MPI, but seems indeed to be very useful.

I think, it would be more in the spirit of MpiFx, if we packed the window object into a special type. The test program could then look like
```
program test_shared_memory
  use libmpifx_module
  implicit none

  type(mpifx_comm) :: globalcomm, nodecomm
  type(mpifx_win) :: win
  integer, parameter :: length = 7
  integer, pointer :: data_pointer(:)

  call mpifx_init()
  call globalcomm%init()

  ! Create a new communicator for all ranks on a node first
  call globalcomm%split_type(MPI_COMM_TYPE_SHARED, globalcomm%rank, nodecomm)

  call win%allocate_shared(nodecomm, length, dataptr)
  call win%lock()

  ! Only rank 0 writes data into the array
  if (win%comm%lead) then
    dataptr(:) = 42
  end if

  call win%sync()
  call win%unlock()

  ! All ranks on the node will read the same value
  write(*, "(2(A,1X,I0,1X))") "ID:", nodecomm%rank, "VALUE:", data_pointer(1)

  call win%free()
  call mpifx_finalize()

end program test_shared_memory
```
The type could store the window ID as well as the communicator for which it was created for to avoid mistakes by passing the wrong one when syncing.

File: lib/mpifx_shared_memory.fpp (hunk 1 above)
Comment (terminationshock) [reply lib/mpifx_shared_memory.fpp]: Thanks for pointing this out. I have changed that. :+1:

**Comment by terminationshock** [comment]
I have updated the PR so that the test program now works exactly as suggested.

**Pull request merged**